### PR TITLE
Indicate a build is in progress in Dashboard build page

### LINKED ISSUE #2542: Indicate a build is in progress in Dashboard build page
When I go to a build in GitHub Actions, I get an animated indication that a build is happening, and how long it's been going for:

Screen Shot 2021-11-29 at 9 55 32 AM

When it's done, the yellow animated circle becomes a green check mark or a red 'x'. 

Screen Shot 2021-11-29 at 9 58 45 AM

Let's do something like this in our build dashboard, so you know what's happening when you get there.  Especially when a build step is not producing any output, it can be hard to know if it's still going.

### PR DESCRIPTION
**Pull request by Yoda-Canada**
Fixes #2542
- [X] An animated indication shows a build is happening.
- [X] The animated indicator shows the percentage being built.
- [x] The indicator shows how long it's been going for.
- [x] When it's done, the yellow animated circle becomes a green checkmark or a red ' X'.

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [X] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Checklist

- [ ] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

**Comment by humphd** [comment]
Post a short video of how it looks for reviewers, please.

**Comment by Yoda-Canada** [comment]
https://vimeo.com/652923172/153ee981a9

**Comment by humphd** [comment]
NOTE: you can drag-and-drop video files into this Issue in a comment, and GitHub will host it for you.

We can't use a % since we don't have any concept of where we are in a build.

**Comment by Yoda-Canada** [comment]
https://user-images.githubusercontent.com/55161917/144630593-96189add-8327-4c58-8dee-9e0e10486881.mp4

**Comment by Yoda-Canada** [comment]
@humphd  Do we have the concept of the time or just our estimate
a388a1182d01aed861346941cf7ceb6

**Comment by humphd** [comment]
We know when a build started (Datetime) and therefore can calculate how long it's been running, so we could do something like `1m 15s`, yes.

**Comment by HyperTHD** [comment]
You need to link the issue you're closing so that github automatically closes it when your PR gets merged.
Please do this ASAP @Yoda-Canada

**Comment by Yoda-Canada** [comment]
Add the datetime calculator function. However it doesn't automatically triggered by build. I will try to do it tomorrow. 
https://user-images.githubusercontent.com/55161917/145139239-b0f856c0-f5f6-4a67-a434-6ffbe8d73acd.mp4

File: src/api/status/public/css/animated-indication.css [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+  height: 30px;
+  color: #e6b800;
+  text-align: center;
+  overflow: hidden;
+}
+.box .counter span {
+  font-size: 10px;
+  line-height: 25px;
+  font-weight: bold;
+}
```
Comment (Andrewnt219) [inline src/api/status/public/css/animated-indication.css:46]: Use an icon from [google material icons](https://fonts.google.com/icons?selected=Material+Icons). Don't have to make your own.

File: src/api/status/public/css/animated-indication.css [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+h8 {
+  color: #000;
+  position: absolute;
+  width: 100%;
+  text-align: right;
+  top: 25%;
+  right: 2%;
+  letter-spacing: 1px;
+}
```
Comment (Andrewnt219) [inline src/api/status/public/css/animated-indication.css:55]: I believe this HTML tag doesn't exist.

File: src/api/status/public/css/animated-indication.css [hunk 3]
```diff
@@ -0,0 +1,74 @@
+* {
+  margin: 0;
+  padding: 0;
+  font-family: sans-serif;
+}
```
Comment (Andrewnt219) [inline src/api/status/public/css/animated-indication.css:5]: This is not necessary.

File: src/api/status/public/js/build-log/datetime-calculator.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+  const diff = currentDate - date;
+  const days = Math.floor(diff / 86400);
+  const hours = Math.floor(diff / 3600) % 24;
+  const minutes = Math.floor(diff / 60) % 60;
+  const seconds = diff % 60;
+
+  document.getElementById(
+    'calculatorData'
+  ).innerHTML = `${days} days, ${hours} h ${minutes} m ${seconds} s`;
```
Comment (Andrewnt219) [inline src/api/status/public/js/build-log/datetime-calculator.js:14]: A build is several hours at most, skip the `day`.

File: src/web/src/components/Posts/GitHubInfo.tsx [hunk 5]
```diff
@@ -57,9 +57,10 @@ const filterGitHubUrls = (urls: string[]) => {
     // Ex: /Seneca-CDOT/telescope ✅
     // Ex: /Seneca-CDOT/telescope/pull/2367/commits/d3fagd3fagd3fagd3fagd3fagd3fag4d41265748 ✅
     // Ex: /Seneca-CDOT/telescope/issues ✅
-    const matches = /^\/(?<user>[^/]+)\/(?<repo>[^/]+)((\/(.*))?\/(?<type>[^/]+)\/(?<id>(\d+|\w{40}))\/?$)?/i.exec(
-      pathname
-    );
+    const matches =
```
Comment (Andrewnt219) [inline src/web/src/components/Posts/GitHubInfo.tsx:60]: Unnecessary changes, revert this whole file.

File: src/web/src/student-quotes.js [hunk 6]
```diff
@@ -25,8 +25,7 @@ export default [
   },
   {
     author: 'Abdulbasid Guled',
-    url:
-      'https://dev.to/hyperthd/4-months-of-open-source-what-i-ve-learned-from-taking-dps909-and-what-i-plan-to-do-about-dps911-next-month-1g62',
+    url: 'https://dev.to/hyperthd/4-months-of-open-source-what-i-ve-learned-from-taking-dps909-and-what-i-plan-to-do-about-dps911-next-month-1g62',
```
Comment (Andrewnt219) [inline src/web/src/student-quotes.js:28]: Unnecessary changes, revert this whole file.

**Review by Andrewnt219 — CHANGES_REQUESTED** [review]
Can you update the branch? Your changes to build.html should go into builds.hbc

**Comment by Yoda-Canada** [comment]
> Can you update the branch? Your changes to build.html should go into builds.hbc
DO you mean I need to merge master? Is it correct to use the following command?
```
git pull upstream master
git checkout issue-2542
git merge master
```

**Comment by Andrewnt219** [comment]
@Yoda-Canada 
```
git checkout your-branch-name
git pull --rebase upstream master
```
Then solve all the conflict during the rebase.

**Comment by AmasiaNalbandian** [comment]
> You need to link the issue you're closing so that github automatically closes it when your PR gets merged. Please do this ASAP @Yoda-Canada

I added the issue in the side manually, sorry about this. For future please use one of the [keywords from GitHub](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) to link the issue to the PR.

File: src/api/status/public/css/animated-indication.css (hunk 2 above)
Comment (Yoda-Canada) [reply src/api/status/public/css/animated-indication.css:55]: it set the 'Staging' position.

File: src/web/src/components/Posts/GitHubInfo.tsx (hunk 5 above)
Comment (Yoda-Canada) [reply src/web/src/components/Posts/GitHubInfo.tsx:60]: when pushing the new commit to GitHub, the system autotest. If I don't change it, it can't be uploaded.

File: src/web/src/student-quotes.js (hunk 6 above)
Comment (Yoda-Canada) [reply src/web/src/student-quotes.js:28]: when pushing the new commit to GitHub, the system autotest. If I don't change it, it can't be uploaded.

Comment (Andrewnt219) [reply src/web/src/student-quotes.js:28]: What do you mean?

Comment (Andrewnt219) [reply src/web/src/student-quotes.js:28]: I worked on the dashboard recently and I didn't have to make changes to these files.

File: src/api/status/public/css/animated-indication.css (hunk 2 above)
Comment (Andrewnt219) [reply src/api/status/public/css/animated-indication.css:55]: What is `h8` tag? Can you provide a link to its documentation?

File: src/web/src/student-quotes.js (hunk 6 above)
Comment (Yoda-Canada) [reply src/web/src/student-quotes.js:28]: no worry, if the issue reproduced next time, I will post screen shot

**Comment by Yoda-Canada** [comment]
> > Can you update the branch? Your changes to build.html should go into builds.hbc
@Andrewnt219 when I change build.html into builds.hbc, webpage can't display the data time calculator. Could you help me to find the issue:
step 1: in the ` src/api/status/public/js/pages/build.js ` file add the following code
 ``` 
            import datetimeCalculator from '../build-log/datetime-calculator.js';
                 window.addEventListener('load', () => {
                 datetimeCalculator();
                setInterval(datetimeCalculator, 1000);
               }
 ```
step 2: in the ` src/api/status/src/views/builds.hbs `  file add the follwing code
   ```
           <div id="calculatorData" style="float: right"></div>
   ```

**Comment by Andrewnt219** [comment]
Can you update your branch first? 

> when I change build.html into builds.hbc, webpage can't display the data time calculator. Could you help me to find the issue

Any screenshot for errors or any error messages will be helpful.

**Comment by Yoda-Canada** [comment]
@humphd I have almost finished this issue, but when I want to commit my code, it displays an error. Could you help me?
image
image

**Comment by Yoda-Canada** [comment]
image

**Comment by Andrewnt219** [comment]
Try rebasing your branch and run `pnpm i` again.

**Comment by Yoda-Canada** [comment]
I followed the instruction of ` https://dev.to/menghif/transition-to-pnpm-4alf` and when I execute the `pnpm install`, there is an error
image

**Comment by Andrewnt219** [comment]
On Window, you should open Powershell and run this script first.
```
Set-ExecutionPolicy -Scope CurrentUser -ExecutionPolicy Unrestricted
```

See: https://stackoverflow.com/questions/41117421/ps1-cannot-be-loaded-because-running-scripts-is-disabled-on-this-system

**Comment by Yoda-Canada** [comment]
@Andrewnt219 can I delete the `src/api/status/public/css/animated-indication.css` directly in PR 2562? I have deleted this file in local, but it still displays in github

**Comment by Yoda-Canada** [comment]
@Andrewnt219 I can't find the already deleted file -` src/api/status/public/css/animated-indication.css `. Can I create a new one, then delete and commit?

**Comment by Yoda-Canada** [comment]
how to fix the End-to-EndN Tests error? It looks like the npm error, and I have deleted the npm before installed pnpm

**Comment by Andrewnt219** [comment]
> @Andrewnt219 I can't find the already deleted file -src/api/status/public/css/animated-indication.css. Can I create a new one, then delete and commit?

Deleting from the PR is fine too I guess.

> how to fix the End-to-EndN Tests error? It looks like the npm error, and I have deleted the npm before installed pnpm

Not sure why it failed either. Just fix ESLint and push again.

**Comment by Yoda-Canada** [comment]
In the file ` src\api\status\public\js\build-log\datetime-calculator.js` , add the following test data in Line7-11:
``` 
  data = {
    "startedAt": "2021-12-16T01:45:57.509Z",
    "stoppedAt": "null",
    "code": 0
  }
```

image

 [New Recording - 12/15/2021, 11:21:59 PM](https://vimeo.com/657281694/fb9b5a53e8)

**Comment by Yoda-Canada** [comment]
In the file ` src\api\status\public\js\build-log\datetime-calculator.js` , add the following test data in Line7-11:
``` 
  data = {
    "startedAt": "2021-12-16T01:45:57.509Z",
    "stoppedAt": "2021-12-16T01:50:59.509Z",
    "code": 1
  }
```
image

**Comment by Yoda-Canada** [comment]
In the file src\api\status\public\js\build-log\datetime-calculator.js , add the following test data in Line7-11:
```
  data = {
    "startedAt": "2021-12-16T01:45:57.509Z",
    "stoppedAt": "2021-12-16T01:50:59.509Z",
    "code": 0
  }
```
image

File: src/api/status/public/js/build-log/datetime-calculator.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
+const buildDateCaltulator = document.getElementById('calculatorData');
+const buildLoading = document.getElementById('build-loading');
+const buildSuccess = document.getElementById('build-success');
+const buildFail = document.getElementById('build-fail');
+
+export default function calculatorBuildInfo(data) {
+  const startedAtValue = data.startedAt;
+  const stoppedAtValue = data.stoppedAt;
+  const resultValue = data.result;
```
Comment (Andrewnt219) [inline src/api/status/public/js/build-log/datetime-calculator.js:9]: use [destructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) here. You can also check why ESLint failed before and the solution for [prefer-destructing](https://eslint.org/docs/rules/prefer-destructuring)

File: src/api/status/public/js/build-log/datetime-calculator.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+  if (startedAtValue !== 'null' && stoppedAtValue === 'null') {
+    const date = Math.abs((new Date(startedAtValue).getTime() / 1000).toFixed(0));
+    const currentDate = Math.abs((new Date().getTime() / 1000).toFixed(0));
+    const diff = currentDate - date;
+    const hours = Math.floor(diff / 3600) % 24;
+    const minutes = Math.floor(diff / 60) % 60;
+    const seconds = diff % 60;
```
Comment (Andrewnt219) [inline src/api/status/public/js/build-log/datetime-calculator.js:17]: This should be extracted into a function. It's used three times here for the same calculation.

File: src/api/status/public/js/build-log/datetime-calculator.js [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    buildDateCaltulator.innerText = `${hours} h ${minutes} m ${seconds} s`;
+    buildFail.hidden = true;
+    buildSuccess.hidden = true;
+    buildLoading.hidden = false;
```
Comment (Andrewnt219) [inline src/api/status/public/js/build-log/datetime-calculator.js:21]: It's easier to have only one icon placeholder in the HTML and use JavaScript to add icon in, rather than managing the display of all three icons.

```js
const buildIcon = document.getElementById("#build-icon");
buildIcon.innerHTML = "motion_photos_on"
buildIcon.classList.add("fa-spin");
```

**Review by Andrewnt219 — CHANGES_REQUESTED** [review]
Glad to see this is taking shape. There is still room for improvement.

**Comment by humphd** [comment]
I was just watching a build on Vercel, and this is how they do it, with a spinner that flips to a check-mark or 'x', and a pretty-printed time value:

Screen Shot 2021-12-16 at 5 05 00 PM
Screen Shot 2021-12-16 at 5 04 17 PM

File: src/api/status/public/js/pages/build.js [hunk 10]
```diff
@@ -1,6 +1,9 @@
 import checkForBuild from '../build-log/check-for-build.js';
+import calculatorBuildInfo from '../build-log/datetime-calculator.js';
 
 window.addEventListener('load', () => {
   checkForBuild();
+  calculatorBuildInfo();
```
[the remark this answers is not in the document]
Comment (Yoda-Canada) [reply src/api/status/public/js/pages/build.js:6]: sure, I delete the ` calculatorBuildInfo();`, and just use the setInterval.

File: src/api/status/src/views/builds.hbs [hunk 11]
```diff
@@ -31,6 +31,24 @@
       </div>
     </div>
   </nav>
+  <div class="container-fluid py-2">
```
[the remark this answers is not in the document]
Comment (Yoda-Canada) [reply src/api/status/src/views/builds.hbs:34]: In the beginning, I want to move this into the build header. But after I studied issue #2416, I think they are different functions, maybe it's better to separate these two items

[the remark this answers is not in the document]
Comment (Andrewnt219) [reply src/api/status/src/views/builds.hbs:34]: Are they? Your result and total duration are for `current` build, and they're the same as those two columns in `build-header`. Data is from the same source, too.

File: src/api/status/public/js/pages/build.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (Andrewnt219) [reply src/api/status/public/js/pages/build.js:6]: Don't have to call calculatorBuildInfo() here, inside `checkForBuild` is enough.

**Comment by Andrewnt219** [comment]
@Yoda-Canada any updates on the current state of this PR?

**Comment by Yoda-Canada** [comment]
@Andrewnt219 I haven't updated the current state yet. I will review it again.

**Comment by tpmai22** [comment]
@Yoda-Canada how is everything for this PR, would you need some help rebasing it ?

**Comment by Yoda-Canada** [comment]
@tpmai22 Most of the issues has been solved, except for the following issue.  Could you help me solve this problem? I really appreciate your assistance.
image

**Comment by tpmai22** [comment]
I would love to take a look and see

**Comment by Andrewnt219** [comment]
@tpmai22 you may want to wait for #2653 to merge.

**Comment by tpmai22** [comment]
@Andrewnt219 for this issue I try to catch up with the conversation between you and @Yoda-Canada  but I'm still a bit lost. Would you mind to shred some light ?

**Comment by Andrewnt219** [comment]
@tpmai22 in #2582, a build header was added to show the info of the ongoing build (see screenshots). At the moment, the header just shows the result "Error" or "Good". It is missing an "In progress" state. 

The current PR is also showing animation/loading indicators of the 3 statuses in addition to just plain text.

**Comment by tpmai22** [comment]
We can close this now refer to #2665
